Automattic/woocommerce-payments
# Add customer facing prices for customer multi-currency

Fixes #1769 

## Changes proposed in this Pull Request

This PR converts frontend prices to the selected currency, maintaining compatibility with WooCommerce's native features. Compatibility with other extensions will be addressed as part of another issue. But this work *should* also be compatible with WooCommerce Subscriptions, barring a few edge cases like Subscriptions-specific coupons.

The way prices are converted and formatted is by using some WooCommerce's hooks. Two classes were created to handle them: [Frontend_Prices](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-5e2b0c63cb5e2acb8f645e5505db93b758eb74c36703584c42367c726bb2ed0bR15) and [Frontend_Currencies](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-c67278bce0bbc54e546375c4d665efd19475a58c00f2586e6cc0f01cb981064aR15). As their names might indicate, they are only loaded in the frontend, by using the [same check as WC](https://github.com/woocommerce/woocommerce/blob/6667233eb11b3d925f5df502e4090742bf1b0c0f/includes/class-woocommerce.php#L359). Using that loading check means that, for now, the multi-currency prices won't work for the checkout blocks. Since we're only building support for blocks in the future, I think it'd be best to address this as part of another issue. Is that okay?

The Frontend_Prices class is responsible for the price conversion for products, shipping rates, and coupons. The Frontend_Currencies one is responsible for using the right symbol, formatting, separators, and decimal places when rendering currencies, and uses locale data from [woocommerce/i18n/locale-info.php](https://github.com/woocommerce/woocommerce/blob/b19500728b4b292562afb65eb3a0c0f50d5859de/i18n/locale-info.php). The data in the locale file can be filtered using [`wcpay_multi_currency_<currency_code>_format`](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-c67278bce0bbc54e546375c4d665efd19475a58c00f2586e6cc0f01cb981064aR126).

I made a few assumptions when calculating and adjusting the prices, such as we'd always want to round up the converted amount, we wouldn't want to apply charm pricing to coupons, and we wouldn't want to ceil and apply charm pricing to taxes. Do they seem right?

The price calculation and adjustment can be found [here](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-b17bf67925ece08a12ad748cd870705e5d668d3fa19214040f4841daaf712cdcR263-R326). When adjusting the price, we always round up. However, PHP doesn't provide a `ceil` method that accepts the number of decimal places and I had to write [`ceil_price`](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-b17bf67925ece08a12ad748cd870705e5d668d3fa19214040f4841daaf712cdcR323). Is that solution robust and reliable enough for our use case?

A few settings are being mocked currently, and can be changed by using filters:
- `wcpay_multi_currency_round_precision`: The number of decimal places to use when rounding, can be negative.
- `wcpay_multi_currency_charm_pricing`: The charm pricing modifier that is added when adjusting the price.
- `wcpay_multi_currency_apply_charm_only_to_products`: Setting this to false indicates that shipping rates should have a charm modifier as well.

A few of the tests introduced in this PR rely on the mocked currencies setup and will have to be modified when we start fetching currencies for real.









## Testing instructions





You can select a different currency by using the `?currency=<code>` URL param when visiting a frontend page.

To be able to test different currency formatting and zero-currencies, I've modified `get_mock_currencies` to include:

```
return [
	[ 'USD', '1.00' ],
	[ 'CHF', '0.89873' ],
	[ 'HUF', '286.33421' ], // Zero decimal currency.
	[ 'KRW', '1127.7053' ], // Zero decimal currency.
	...
```

We need to make sure that the select currency is working with the following WooCommerce's features:

- Price formatting and conversion
    - Load up the Shop page with the default store currency and check that all prices look right
    - Select a different currency and ensure that the prices have been converted and formatted accordingly:
        - `USD` should look like `$x'xxx.xx`
        - `CHF` should look like `CHF x,xxx.xx`
        - `HUF` should look like `x xxx Ft`
        - `KRW` should look like `xx,xxx₩`
    - Make sure that discounted products and variation price ranges are rendered correctly
- Simple products
    - Make sure their price is converted in the Shop, Product, Cart, and Checkout pages
    - Purchase the product and check that the order has the right currency and price
    - Navigate to the transactions page and ensure that the currency used was the selected one
- Variable products
    - Make sure their price is converted in the Shop, Product, Cart, and Checkout pages
    - Modify the selected currency and ensure the price range was updated. It is cached by WC by default but should be refreshed when the exchange rate changes
    - Purchase the product and check that the order has the right currency and price
    - Navigate to the transactions page and ensure that the currency used was the selected one
- Coupons
    - Enable coupon codes in the General WooCommerce settings
    - Create three coupons, one for each type
    - Ensure that the fixed cart discount and fixed product discount amounts are converted when using them
    - Ensure that the percentage discount works correctly, without having its amount converted
    - Add a minimum and maximum spend usage restriction to a coupon, and check that those values are converted and matched against the cart subtotal
- Shipping
    - Add a free and a flat rate shipping method to your shipping zone
    - Make sure that they work correctly on the checkout page
        - If `apply_charm_only_to_products` is enabled, the shipping rate should only be ceiled after conversion
        - If it's not, the shipping rate should also have the charm pricing added to it
- Taxes
    - Enable taxes in the General WooCommerce settings and create a standard tax rate
    - Ensure that taxes are correctly applied and converted when using both "Prices inclusive of tax" and "Prices exclusive of tax"
    - Ensure that the shipping methods' taxes are also converted properly under both configurations


-------------------
